pkg/volumes/alicloud/volumes.go

```diff
@@ -0,0 +1,307 @@
+		case "Attaching":
+			glog.V(2).Infof("Waiting for volume %q to be attached (currently %q)", volumeID, v.Status)
+			// continue looping
+
+		default:
+			return fmt.Errorf("Observed unexpected volume state %q", v.Status)
+		}
+
+		time.Sleep(10 * time.Second)
+	}
+}
+
+func (a *AlicloudVolumes) FindVolumes() ([]*volumes.Volume, error) {
+	return a.findVolumes(true)
+}
+
+func (a *AlicloudVolumes) findVolumes(filterByAZ bool) ([]*volumes.Volume, error) {
+	request := ecs.CreateDescribeDisksRequest()
+	request.RegionId = a.region
+
+	if filterByAZ {
+		request.ZoneId = a.az
+	}
+
+	var tags []ecs.DescribeDisksTag
+
+	for _, key := range a.matchTagKeys {
+		tags = append(tags, ecs.DescribeDisksTag{
+			Key:   key,
+			Value: "",
+		})
+	}
+
+	for key, value := range a.matchTags {
+		tags = append(tags, ecs.DescribeDisksTag{
+			Key:   key,
+			Value: value,
+		})
+	}
+
+	request.Tag = &tags
+	request.PageSize = requests.NewInteger(PageSizeLarge)
+	request.PageNumber = requests.NewInteger(1)
+
+	return a.describeVolumes(request)
+}
+
+func (a *AlicloudVolumes) describeVolumes(request *ecs.DescribeDisksRequest) ([]*volumes.Volume, error) {
+	var found []*volumes.Volume
+
+	for {
+		response, err := a.ecs.DescribeDisks(request)
+		if err != nil {
+			return nil, err
+		}
+
+		if response == nil || len(response.Disks.Disk) < 1 {
+			break
+		}
+
+		disks := response.Disks.Disk
+
+		for _, disk := range disks {
+			etcdName := disk.DiskId
+
+			if a.nameTag != "" {
+				for _, t := range disk.Tags.Tag {
+					if a.nameTag == t.TagKey {
+						v := t.TagValue
+						if v != "" {
+							tokens := strings.SplitN(v, "/", 2)
+							etcdName = a.clusterName + "-" + tokens[0]
+						}
+					}
+				}
+			}
+
+			vol := &volumes.Volume{
+				MountName:  "master-" + disk.DiskId,
+				ProviderID: disk.DiskId,
+				EtcdName:   etcdName,
+				Info: volumes.VolumeInfo{
+					Description: disk.Description,
+				},
+				Status:     disk.Status,
+				AttachedTo: disk.InstanceId,
+			}
+
+			if vol.AttachedTo == a.instanceID {
+				vol.LocalDevice = disk.Device
+			}
+
+			found = append(found, vol)
+		}
+
+		if len(response.Disks.Disk) < PageSizeLarge {
+			break
+		}
+
+		nextPage, err := getNextpageNumber(request.PageNumber)
+		if err != nil {
+			return nil, err
+		}
+		request.PageNumber = nextPage
+	}
+
+	return found, nil
+}
+
+func (a *AlicloudVolumes) MyIP() (string, error) {
+	return a.localIP, nil
+}
+
+func getNextpageNumber(number requests.Integer) (requests.Integer, error) {
+	page, err := strconv.Atoi(string(number))
+	if err != nil {
+		return "", err
+	}
+	return requests.NewInteger(page + 1), nil
+}
+
+// FindMountedVolume implements Volumes::FindMountedVolume
+func (a *AlicloudVolumes) FindMountedVolume(volume *volumes.Volume) (string, error) {
+	device := volume.LocalDevice
+
+	_, err := os.Stat(volumes.PathFor(device))
+	if err == nil {
+		return device, nil
+	}
+	if os.IsNotExist(err) {
+		if strings.HasPrefix(device, "/dev/xvd") {
```

Comment:
looks dirty for me, what is purpose of these things?

Reply:
You are right, this is very dirty. But the reality is that when check one disk with a device name: `/dev/xvdb` in Alicloud web console, it is attach to the OS as `/dev/vdb`. This is also very strange to me. I have no idea why Alicloud does this kind of strange thing. It also took me a while to debug this. 😫

Reply:
I think it's because the name in the OS is determined by the OS, not by the hypervisor and thus not by the cloud.  AWS has some similar stuff, where we have a list of paths we try.  It's just the way it is, as far as I can tell.

If you can see the disk by volume name (e.g. in `/dev/disk/by-id`) that is much more reliable.  GCE and the NVME disks on AWS do present in this way.  Definitely worth checking if this does work on Alicloud!  But otherwise, this logic looks right!

Reply:
@justinsb, I just checked on an Alicloud instance, under `/dev/disk/`, there are only `by-path` and `by-uuid`. 

I'm also reaching out to Alicloud engineers for more information.

Reply:
Got a reply from Alicloud engineers. Currently we can't see disk by volume name/id from OS.

Reply:
Another reply from Alicloud side, this `/dev/disk/by-id` feature is currently under development, and will be released at the end of Jan 2020.

I will submit another PR for this when this feature is live. @justinsb